code/mission/missionparse.cpp

```diff
@@ -5801,6 +5801,87 @@ void parse_sexp_containers()
 		stuff_sexp_map_containers();
 		required_string("$End Maps");
 	}
+
+	// jg18 - persistence-related checking
+	// adapted from parse_variables()
+
+	// do this stuff only when playing through a campaign
+	if (Fred_running || !(Game_mode & GM_CAMPAIGN_MODE)) {
+		return;
+	}
+
+	// first update this mission's containers from campaign-persistent containers
+	for (const auto &current_pc : Campaign.persistent_containers) {
+		auto *p_container = get_sexp_container(current_pc.container_name.c_str());
+		if (p_container != nullptr) {
+			auto &container = *p_container;
+
+			// if this is an eternal container that shares the same name as a non-eternal, warn but do nothing
+			if (container.is_eternal()) {
+				error_display(0,
+					"SEXP container %s is marked eternal but has the same name as another persistent container. One of "
+					"these should be renamed to avoid confusion",
+					container.container_name.c_str());
+			} else if (container.is_persistent()) {
+				if (container.type_matches(current_pc)) {
```

Comment:
See `sexp_container.cpp` for definition of `type_matches()`, but it's comparing only the non-persistence-related container type flags.

My reasoning for doing type matching in the first place was that replacing (i.e., updating) persistent containers shouldn't replace, for example:
- a map container with a list container
- a container using string data with a container using numeric data
- a map using string keys with a map using numeric keys

If/when strict typing is added, we should take that into account, too.

But @Karajorma what do you think should be the expected behavior here?

I don't think replacing persistent variables checks variable types, but type differences matter a lot more with containers.